server/src/main/java/org/elasticsearch/ingest/IngestSourceAndMetadata.java

```diff
@@ -0,0 +1,588 @@
+        }
+        Map<String, Object> metadata = Maps.newHashMapWithExpectedSize(IngestDocument.Metadata.values().length);
+        Map<String, Object> source = new HashMap<>(sourceAndMetadata);
+        for (String metadataName : VALIDATORS.keySet()) {
+            if (sourceAndMetadata.containsKey(metadataName)) {
+                metadata.put(metadataName, source.remove(metadataName));
+            }
+        }
+        return new Tuple<>(source, metadata);
+    }
+
+    /**
+     * Fetch the timestamp from the ingestMetadata, if it exists
+     * @return the timestamp for the document or null
+     */
+    public static ZonedDateTime getTimestamp(Map<String, Object> ingestMetadata) {
+        if (ingestMetadata == null) {
+            return null;
+        }
+        if (ingestMetadata.get(IngestDocument.TIMESTAMP)instanceof ZonedDateTime timestamp) {
+            return timestamp;
+        }
+        return null;
+    }
+
+    /**
+     * get the source map, if externally modified then the guarantees of this class are not enforced
+     */
+    public Map<String, Object> getSource() {
+        return source;
+    }
+
+    /**
+     * get the metadata map, if externally modified then the guarantees of this class are not enforced
+     */
+    public Map<String, Object> getMetadata() {
+        return metadata;
+    }
+
+    // These are available to scripts
+    public String getIndex() {
+        return getString(IngestDocument.Metadata.INDEX.getFieldName());
+    }
+
+    public void setIndex(String index) {
+        put(IngestDocument.Metadata.INDEX.getFieldName(), index);
+    }
+
+    public String getId() {
+        return getString(IngestDocument.Metadata.ID.getFieldName());
+    }
+
+    public void setId(String id) {
+        put(IngestDocument.Metadata.ID.getFieldName(), id);
+    }
+
+    public String getRouting() {
+        return getString(IngestDocument.Metadata.ROUTING.getFieldName());
+    }
+
+    public void setRouting(String routing) {
+        put(IngestDocument.Metadata.ROUTING.getFieldName(), routing);
+    }
+
+    public String getVersionType() {
+        return getString(IngestDocument.Metadata.VERSION_TYPE.getFieldName());
+    }
+
+    public void setVersionType(String versionType) {
+        put(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), versionType);
+    }
+
+    public long getVersion() {
+        Number version = getNumber(IngestDocument.Metadata.VERSION.getFieldName());
+        assert version != null : IngestDocument.Metadata.VERSION.getFieldName() + " validation allowed null version";
+        return version.longValue();
+    }
+
+    public void setVersion(long version) {
+        put(IngestDocument.Metadata.VERSION.getFieldName(), version);
+    }
+
+    // timestamp isn't backed by the map
+    public ZonedDateTime getTimestamp() {
+        return timestamp;
+    }
+
+    // These are not available to scripts
```

Comment:
If they are not available to scripts, why are they in the metadata map? It doesn't appear to be added when the IngestSourceAndMetadata is created, so when and how does it get in there?